pkg/controller/direct/dataproc/job_controller.go

```diff
@@ -0,0 +1,498 @@
+	desired     *krm.DataprocJob // Desired state
+	actual      *pb.Job          // Actual state fetched from GCP
+	generatedId *string
+}
+
+func (a *dataprocJobAdapter) Find(ctx context.Context) (bool, error) {
+	if a.id.ID() == "" {
+		// If JobID is not known (e.g., before creation), we can't GET the job.
+		// However, we might try to find it using a list call with a filter if
+		// the user provided a stable ID via placement.jobId or resourceID annotation.
+		// For now, assume we can only find by service-generated ID.
+		// TODO(kcc): Implement List+Filter based on user-provided ID if available.
+		klog.V(2).Infof("cannot find dataproc job without a job ID (project %s, region %s)", a.id.Parent().ProjectID, a.id.Parent().Location)
+		return false, nil
+	}
+
+	req := &pb.GetJobRequest{
+		ProjectId: a.id.Parent().ProjectID,
+		Region:    a.id.Parent().Location,
+		JobId:     direct.ValueOf(a.generatedId),
+	}
+
+	klog.V(2).Infof("getting dataproc job %q", a.id)
+	actual, err := a.gcpClient.GetJob(ctx, req)
+	if err != nil {
+		if direct.IsNotFound(err) {
+			klog.V(2).Infof("dataproc job %q not found", a.id)
+			return false, nil
+		}
+		return false, fmt.Errorf("getting dataproc job %q: %w", a.id, err)
+	}
+	klog.V(2).Infof("found dataproc job %q", a.id)
+
+	a.actual = actual
+
+	// Check if the job is terminated (completed, cancelled, errored).
+	// Once terminated, it's effectively immutable and shouldn't be reconciled further.
+	// We treat it as "found" but signal it shouldn't be updated/deleted by KCC if it's done.
+	// KCC deletion should still work if the user deletes the KRM object.
+	// if a.actual.Status != nil && (a.actual.Status.State == pb.JobStatus_DONE || a.actual.Status.State == pb.JobStatus_CANCELLED || a.actual.Status.State == pb.JobStatus_ERROR) {
+	// 	klog.Infof("Dataproc job %s/%s/%s is already in a terminal state (%s).", a.id.ProjectID, a.id.Region, a.id.JobID, a.actual.Status.State)
+	// 	// TODO: How to signal this state back to the controller logic? Maybe via status?
+	// }
+
+	return true, nil
+}
+
+func (a *dataprocJobAdapter) Create(ctx context.Context, createOp *directbase.CreateOperation) error {
+	klog.V(2).Infof("creating dataproc job in project %s region %s", a.id.Parent().ProjectID, a.id.Parent().Location)
+
+	mapCtx := &direct.MapContext{}
+	desiredProto := DataprocJobSpec_ToProto(mapCtx, &a.desired.Spec)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+	// We need the full Job message, not just the spec part.
+	desired := &pb.Job{
+		Placement:              desiredProto.Placement,
+		JobUuid:                desiredProto.JobUuid,
+		Scheduling:             desiredProto.Scheduling,
+		YarnApplications:       desiredProto.YarnApplications,
+		DriverSchedulingConfig: desiredProto.DriverSchedulingConfig,
+		Done:                   desiredProto.Done,
+		Labels:                 desiredProto.Labels,
+	}
+	// Populate the job type specific field
+	if a.desired.Spec.HadoopJob != nil {
```

Comment:
I am a bit confused. These conversion should be handled by mapping functions? I would expect `DataprocJobSpec_ToProto` to handle this.